poh/src/poh_recorder.rs

```diff
@@ -977,13 +977,44 @@ impl PohRecorder {
         self.clear_bank();
     }
 
+    fn skew_block_producer_time_nanos(
+        parent_slot: Slot,
+        parent_time: u64,
+        working_bank_slot: Slot,
+        working_bank_time: u64,
+    ) -> u64 {
+        let default_ns_per_slot = DEFAULT_MS_PER_SLOT * 1_000_000;
+        let diff_slots = working_bank_slot.saturating_sub(parent_slot);
+
+        let min_working_bank_time = parent_time.saturating_add(1);
+        let max_working_bank_time =
+            parent_time.saturating_add(2 * diff_slots * default_ns_per_slot);
+
+        working_bank_time
+            .max(min_working_bank_time)
+            .min(max_working_bank_time)
+    }
+
     pub fn working_bank_block_producer_time_nanos(&self) -> u64 {
```

Comment:
this value also needs to go into the clock sysvar right? seems we should do it in bank.freeze

Reply:
2 things:

We only send out the footer once the bank is frozen. Once a bank is frozen no more account updates should be made (the state of the bank is "frozen").

There is no "leader replay"
https://github.com/anza-xyz/alpenglow/blob/4bfe9ed011926f73f6cda04b00ceb07cfce881a6/core/src/replay_stage.rs#L3464-L3466
The transactions are recorded into the bank directly by banking stage https://github.com/anza-xyz/alpenglow/blob/4bfe9ed011926f73f6cda04b00ceb07cfce881a6/core/src/banking_stage/consumer.rs#L174

This is why a bank can become frozen before it has been done shredding